### DISCUSSION TITLE
Named "convention" for embed with "extra classes"

### DISCUSSION THREAD

**Discussion by cavasinf** [kevinpapst/TablerBundle#109] [Q&A] [answered]
Hey @kevinpapst.

I'm in a dead-end with how the modal embed is done.
I want to **add multiple classes** to `modal-body` but we can't.
The only way that is open to devs, is to use `modal_body_attributes` with `style` attribute (ugly ! ⚠️).

I was about to make a PR to solve this but there are some questions:
### How to handle more classes than default `Bootstrap` or `Tabler` classes?
_eg: I want to add more classes to `modal-body` like `p-0`_
1. ```twig
    <div class="modal-body {{bodyExtraClass}}">
       {% block modal_body %}{% endblock %}
    </div>
    ```
2. ```twig
    {% set _bodyClasses = "modal-body " + bodyExtraClass %}

    <div class="{{ _bodyClasses }}">
       {% block modal_body %}{% endblock %}
    </div>
    ```

### Depend of your answer to above but do we name Extra classes:
1. `bodyExtraClass`
2. `bodyClass`

### Does `bodyClass` override default classes?
IMO, I will not understand if I use `bodyClass` option, I'll loose the default `modal-body`.. Kinda weird to me.
But that mean that we cannot remove the default `modal-body` 🤔 

`bodyExtraClass` is good to understand, but maybe too long to write..

### Block VS variables
I'm not a huge fan of "floating strings" inside block to defined parameters..
Like: `modal_size` or `modal_body_attributes `
```diff
{% embed '@Tabler/embeds/modal.html.twig' %}
+   {% block modal_size %}modal-xl{% endblock %}
    {% block modal_title %}
        {{ 'action.edit_interlocutors'|trans }}
    {% endblock %}
+   {% block modal_body_attributes %}
+      {{ parent() }}
+       style="padding:0; margin-bottom: 1rem; border-bottom: 1px solid rgba(98, 105, 118, 0.16)"
+  {% endblock %}
{% endembed %}
```

Is that OK to uses:
1. `Options param` like macros + `blocks` to template
```twig
{% embed '@Tabler/embeds/modal.html.twig' with {
  modalSize : 'xl',
  modalTitle : 'action.edit_interlocutors'|trans,
  modalBodyAttributes : 'style="padding:0; margin-bottom: 1rem; border-bottom: 1px solid rgba(98, 105, 118, 0.16)"'
} %}
   {% block modal_body %}
      <h1>Here's some common HTML!</h1>
  {% endblock %}
{% endembed %}
```
2. Or only `blocks`
```twig
{% embed '@Tabler/embeds/modal.html.twig' %}
    {% block modal_size %}modal-xl{% endblock %}
    {% block modal_title %}
        {{ 'action.edit_interlocutors'|trans }}
    {% endblock %}
    {% block modal_body_attributes %}
       {{ parent() }}
        style="padding:0; margin-bottom: 1rem; border-bottom: 1px solid rgba(98, 105, 118, 0.16)"
   {% endblock %}
{% endembed %}
```

**Comment by kevinpapst** [comment]
Wow, that must have taken ages to create, well explained 👍 

I think an embed works with blocks and a macro works with parameter.

Is something wrong with adding a simple block:
```twig
            <div {% block modal_body_attributes %}{% endblock %} class="modal-body {% block modal_body_class %}{% endblock %}">
                {% block modal_body %}{% endblock %}
            </div>
```

> But that mean that we cannot remove the default modal-body 🤔

Fine with me. If you have a very special use-case, you can always use plain html with the bootstrap classes.

**Reply by cavasinf** [reply]
> Wow, that must have taken ages to create, well explained 👍

20/30 minutes maybe 😄 

-------------------

> I think an embed works with blocks and a macro works with parameter.

Ok, I'll follow that way.

----------------------

> Is something wrong with adding a simple block:

I'm ok with blocks, but does that syntax works inside string ?
```twig
<div class="modal-body {% block modal_body_class %}{% endblock %}"></div>
```

We can use `block()` twig function, 
but I stumble with the [Symfony plugin](https://plugins.jetbrains.com/plugin/7219-symfony-support):
- Overriding `{% block %}` works, it gives me auto-completion with the block name.
- But overriding `block()` doesn't work very well, sometimes auto-completion works and sometimes not..
So I am not in good agreement with it..

--------------
> 
> > But that mean that we cannot remove the default modal-body 🤔
> 
> Fine with me. If you a very special use-case you can always use plain bootstrap.

Good to me 👍

**Reply by kevinpapst** [reply]
Sometimes block() is the most convenient way to create the embed. But if you find places which can be re-written & improved: please do so! This bundle is really meant to make our lives easier.

**Comment by cavasinf** [comment, marked as answer]
To summarize:

1. How to handle more classes than default Bootstrap or Tabler classes?
```twig
<div class="modal-body {% block modal_body_class %}{% endblock %}"></div>
```

2. Depend of your answer to above but do we name Extra classes:
`xxxxx_class`

3. Does bodyClass override default classes?
No ❌ 

4. Block VS variables
`Blocks` inside templates AND `variables` for macros

**Reply by kevinpapst** [reply]
1. ✔️ 
2. I didn't think it through entirely and it my be inconsistent in the existing files... but from what I remember my idea was: empty blocks are called `xxx_class` and `yyy_classes` if they already contain pre-set classes, which can be overwritten.
3. What do you mean?
4. That's how I used them. Not sure if there is some kind of "best practice" out there. It is really just something I am used to.

**Reply by cavasinf** [reply]
1. ✔️ 

2. I've never seen `yyy_classes` in TablerBundle, maybe for AdminLteBundle but not there.. 
But good idea 👍 


3. AS you answer in "point 2", singular variable/block should not override default classes.
If it does, variable name will be in plural.
> cavasinf said:
> Does bodyClass override default classes?

`bodyClass` (singular) doesn't override default classes.

4. [Card embed](https://github.com/kevinpapst/TablerBundle/blob/main/templates/embeds/card.html.twig) is the most chaotic template but the more complete! 
Deeper dive into that for an example:

# Blocks  VS variables
_I'm looking at the first `<div>` ([L18](https://github.com/kevinpapst/TablerBundle/blob/908548e4d6ac0ed94eb3e0ab75364c8a76111e1b/templates/embeds/card.html.twig#L18))_
We have a mix of everything in there:
- Variables : 
   1. `_id`
   2. `_margin_bottom ` 
- Blocks:
   1. block_class
   2. block_attributes

> kevinpapst said:
> Is something wrong with adding a simple block

It's not the case here.. 🤐 

# Blocks usage: `box_class` VS `box_title`
## `box_class`
**Has no needs to be HTML (why block then ?)**
#### Pros:
- Auto-completion ✔️
![image](https://user-images.githubusercontent.com/25293190/190151630-4b2d9c2a-ba12-4d59-b643-3b79cf23602a.png)

####  Cons:
- Condition is really long.......
```twig
{% if block('box_class') is defined %} {{ block('box_class') }}{% endif %}
```

## `box_title`
**Can be HTML**
#### Pros:
- Is contained in variable AND conditional (if not defined)
```twig
{% set _title = block('box_title') is defined ? block('box_title') : null %}
```
- Short to use in templating
#### Cons:
- No Auto-completion ❌ 
![image](https://user-images.githubusercontent.com/25293190/190151667-4d79798a-837e-4346-8eda-69b02b278be0.png)

**Reply by kevinpapst** [reply]
I totally agree with "the most chaotic"!

The history of the embed is terribly old. It came from the original AdminBundle, which I forked to AdminLTeBundle, then used for a couple of years and the re-used in TablerBundle. And I agree that it would need a refactor - but I am not going to do that. Switching my app from AdminLTE to Tabler was a pain and this embed is used literally on every page with all kind of usage combinations.

We can add a "better card" embed and/or try to create wrapper, but changing the current setup would mean hours of work due to BC breaks.

Regarding the unseen "yyy_classes" blocks: yep, you are right and I am wrong, seems I should check the bundle before writing publicly ... but I still like the idea 😁

**Reply by cavasinf** [reply]
> We can add a "better card" embed and/or try to create wrapper

_"i know what i have to do but i don't know if i have the strength to do it" - Star wars_

----------------

Final words..
Embeds should not have variables ? (not like card embed)
I'm not 100% convinced by that 😟 

Back to `modal embed` for `modal_size` example:
```twig
<div class="modal-dialog {% block modal_size %}modal-lg{% endblock %}">
```

1. Why Using block for a class string ?
2. Why Have to set `modal-sm` or `modal-md` as block instead of `sm` or `md`
Reason : Because block will be ugly? 😉 
```twig
<div class="modal-dialog modal-{% block modal_size %}lg{% endblock %}">
```
3. Why not simply use a variable to be more simple:
```twig
{% set _modalSize = "modal-" ~ modalSize|default("lg") %}

<div class="modal-dialog {{ _modalSize }}">
```
4. In a more complex template, it is hard to find which parameter is used to customize that embed.
But when everything is at the top of the file inside parameter, it's more fluid to read/find what you are looking for.
https://github.com/kevinpapst/TablerBundle/blob/908548e4d6ac0ed94eb3e0ab75364c8a76111e1b/templates/components/accordion.html.twig#L1-L10

---------------

Huge point for blocks is that they have auto-completion if used directly in HTML, not like variables.
BUT IMO: 
- blocks are a good way to defined that this "variable" can contain HTML
- variables are for strings/arrays/... NOT HTML

I know we can do something like this:
```twig
{% block modal_size "modal-xl" %}
```
But I really don't like it! My dead ass brain 🧠  keep looking for the `endblock` tag..

**Comment by kevinpapst** [comment]
The last example is really ugly, I never used that.

For 2 & 3 I would say: less flexible. You could use your own class for the sizing `{% block modal_size %}mysize{% end block %}` or even "hack" another class into it `{% block modal_size %}modal-sm myclass{% end block %}`.

I am really not an evangelist who preaches a way of doing something in a certain way, just trying to find a good solution that is understandable to bundle users. It is just my personal feeling, that it is counter-intuitive to pass variables to an embed. 
But for me at the end it is about solving an requirement^^

If you find your force 😁 please post a rewritten embed and I will give it a try!

**Reply by cavasinf** [reply]
> You could use your own class for the sizing

Good point if your custom class is not prefixed by `modal-`.

> I am really not an evangelist who preaches a way of doing something in a certain way

Me neither, just here to discuss/exchange about what is the "better" way to do things 👍 

> it is counter-intuitive to pass variables to an embed

`embed` is like `extend` but with block in it.
`embed` is also like `include` but with blocks.
Both of them allows variables.

> If you find your force 😁 please post a rewritten embed and I will give it a try!

Maybe when I have more time 😄 or holidays ? 🤔
